MaxText/checkpointing.py

```diff
@@ -421,3 +384,72 @@ def save_params_to_path(checkpoint_dir, params, use_ocdbt=True, use_zarr3=True):
   orbax_checkpointer = ocp.PyTreeCheckpointer(use_ocdbt=use_ocdbt, use_zarr3=use_zarr3)
   orbax_checkpointer.save(checkpoint_dir, {"params": params}, force=True)
   print(f"Quantized params checkpoint saved at: {checkpoint_dir}")
+
+
+def maybe_save_checkpoint(checkpoint_manager, state, config, data_iterator, step=None):
+  """Save checkpoint if checkpointing is enabled."""
+  if checkpoint_manager is None:
+    return
+
+  # Determine the effective step for saving a checkpoint.
+  # If 'step' is not provided, this call is for a potential final checkpoint
+  # and use the last completed step from the state.
+  actual_step = (int(state.step) - 1) if step is None else int(step)
+
+  # Determine if a checkpoint save should be forced, overriding the usual `config.checkpoint_period` logic.
+  # This occurs if this function was called:
+  # without an explicit 'step' (implying it's a checkpoint save for final step),
+  # AND the 'actual_step' is a valid step,
+  # AND it's not a step that would normally trigger a checkpoint save.
+  force_ckpt_save = step is None and actual_step != -1 and (actual_step % config.checkpoint_period != 0)
```

Comment:
say training step count = 10
checkpoint freq = 5
then we only save checkpoints at step 0 and 5 because we do % operation

do we want to incorporate the functionality that it saves the checkpoint at step 9?